### ISSUE TITLE
DirectGpuTensorEngine: non-deterministic FP reductions across runs at identical seed (atomic-add scatter in NN gradient kernels)

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet.Tensors#382]
## Summary

`DirectGpuTensorEngine` produces **bit-different floating-point results** across runs at identical seed + identical input + identical model code. On `CpuEngine` the same code produces bit-identical results (verified in AiDotNet PR #1381 / issue #1383 via the reproducer at `tests/AiDotNet.Tests/IntegrationTests/Engines/GpuTransformerDeterminismTests.cs`).

This is **distinct from AiDotNet#1383** (RNG state leakage in `LayerBase.DefaultStrategy`), which was a logic bug affecting both CPU and GPU. The remaining flake — visible after #1383 was fixed — is GPU-side floating-point reduction-order non-determinism inherent to kernels that use atomic accumulation.

## Empirical signature

HarmonicEngine consumer reproducer `Phase_PAPER_A_PathB_Int8_Sanity` after #1383 was applied:

```
Engine: DirectGpuTensorEngine (OpenCL gfx1012:xnack-)
Run 1: FP32 top1 = 6.50%   (delta to INT8 = -6.50pp → FAIL)
Run 2: FP32 top1 = 2.50%   (delta to INT8 = -2.50pp → PASS)
Run 3: FP32 top1 = 2.00%   (delta to INT8 = -2.00pp → PASS)
```

Same source code, same seed, same data — only the engine state across processes differs.

When the same test forces `AiDotNetEngine.ResetToCpu()` at entry, all 4 consecutive runs produce **bit-identical** results (FP32 top1 = 19.00% / INT8 top1 = 19.00% / ppl_ratio = 1.011×).

## Concrete sources in the code

Reduction kernels using `atomicAdd` / `atomic_add_float` for scatter accumulation. Atomic-add reorder is hardware-dependent (work-item scheduling, atomic queue arbitration) and produces non-deterministic FP accumulation order:

**CUDA kernels** (`src/AiDotNet.Tensors/Engines/DirectGpu/CUDA/Kernels/`):
- `CudaNeuralNetKernels.cs:1554` — `atomicAdd(&gradEmbedding[idx * embeddingDim + d], gradOutput[i * embeddingDim + d])` — embedding gradient scatter
- `CudaNeuralNetKernels.cs:2045-2046` — `atomicAdd(&gradGamma[c], ...)` / `atomicAdd(&gradBeta[c], ...)` — LayerNorm gradient reduction
- `CudaNeuralNetKernels.cs:2050` — `atomicAdd(&sumDy[groupIdx], dyGam)` — GroupNorm scatter
- `CudaNeuralNetKernels.cs:1904, 1921, 1977` — generic scatter-add for indexed gather/scatter operations

**OpenCL kernels** (`src/AiDotNet.Tensors/Engines/DirectGpu/OpenCL/Kernels/`):
- `ActivationKernels.cs:1185` — `inline void atomic_add_float(__global volatile float* addr, float val)` — emulated float atomic-add via CAS loop (CAS order is non-deterministic)
- `ActivationKernels.cs:1198-1199` — `atomic_add_float(&output[targetRow * featureSize + col], source[idx])` + `atomic_add(&counts[targetRow], 1)` — scatter-mean accumulation
- `AttentionKernels.cs:26` — attention-specific scatter-add

Each of these accumulates contributions from multiple work-items into a single destination cell. The **order** in which contributions land is determined by the OpenCL/CUDA runtime's work-group scheduler, which is not guaranteed to be deterministic across runs. Floating-point addition is not associative, so different orders → different bit patterns → different gradients → different trained models.

For a Transformer training loop, these non-deterministic gradients compound across steps. At small fixture scale (10KB byte-LM on V=256), the divergence pushes the model into qualitatively different convergence regimes (top-1 oscillates 2%-6.5% on the same seed). At larger scales the relative variance shrinks but doesn't disappear.

## Why this matters

- **Same-seed comparisons are broken**: any A/B / arm-vs-arm test that compares two predictors trained at the same seed gets noise from the engine, not from the variable under test. PyTorch / TF treat this as a known limitation and ship `torch.use_deterministic_algorithms(True)` / `tf.config.experimental.enable_op_determinism()` opt-in flags.
- **Multi-seed sweeps are noisier than they should be**: a "seed sweep" intended to characterize architecture variance picks up engine variance on top.
- **Test envelopes are loose by necessity**: downstream sanity tests have to allow wider quality envelopes to accommodate the engine noise floor, which masks real regressions.

### Path A (RECOMMENDED): Add an opt-in deterministic-engine flag

Mirror the PyTorch `use_deterministic_algorithms(True)` API. Add `AiDotNetEngine.SetDeterministicMode(bool)`. When enabled:
- Replace atomic-add scatter kernels with two-phase **deterministic-reduce kernels** (sort-by-destination + segmented prefix-sum, or atomic-free tree-reduction). Slower (typically 1.3-2× for the affected ops) but **bit-reproducible**.
- Disable any other non-deterministic GPU primitives (CLBlast non-deterministic variants if any, GpuRng if seeded variant is available).
- Document the performance cost and the use case (reproducibility for testing / debugging / paper-grade experiments).

### Path B: Force CPU fallback for non-deterministic ops when seeded RNG is active

Cheaper to implement: when the model has `architecture.RandomSeed.HasValue`, route the affected gradient-scatter ops to CPU even if the engine is otherwise GPU. Catches the most common test-determinism case (sanity / regression tests pin seeds; production runs don't). Costs a per-step CPU round-trip on small ops (~10-50µs), negligible vs the win.

### Path C: Document as known limitation; don't fix

Add a `<para>Reproducibility</para>` section to `DirectGpuTensorEngine` XML doc explaining the limitation and pointing users at `AiDotNetEngine.ResetToCpu()` for bit-reproducible workflows. This is the PyTorch-pre-1.7 stance ("just be aware") — not great but defensible. Consumers (HarmonicEngine, etc.) would need to use `ResetToCpu()` in their own sanity tests as a workaround.

## Reproducer

`AiDotNet/tests/AiDotNet.Tests/IntegrationTests/Engines/GpuTransformerDeterminismTests.cs` from PR #1381 — currently PASSES on `CpuEngine`. To reproduce the GPU bug:
1. Ensure DirectGpuTensorEngine is the active engine (the AiDotNetEngine auto-detection enables it when supported hardware is present)
2. Run the test 4× — bit-identical L2 norms on CPU, varying L2 norms on GPU

## Related issues

- AiDotNet#1380 — closed by PR #1381 (Vector→Tensor bridge in CalculateGradient)
- AiDotNet#1382 — closed by PR #1381 (TransformerArchitecture layers:/numEncoderLayers validator)
- AiDotNet#1383 — closed by PR #1381 (LayerBase.DefaultStrategy + DropoutLayer + LayerHelper.Wire + FeedForwardLayer.SimdRandom seed propagation)
- **This issue** — the remaining flake after the three logical bugs above are fixed; an inherent property of the current GPU kernel implementations
